source/agora/network/DNSResolver.d

```diff
@@ -0,0 +1,263 @@
+
+    /***************************************************************************
+
+        Instantiate a new object of this type
+
+        Params:
+          peers = The 'seed' peers, which are the first point of contact
+                  of the resolver.
+
+    ***************************************************************************/
+
+    public this (const Address[] peers = DefaultDNS)
+    {
+        this.log = Log.lookup(__MODULE__);
+        this.log.enableConsole();
+
+        this.resolvers.length = peers.length;
+        // "Connect" to each of the resolvers
+        // UDP is not connection-oriented, but this binds the socket to the
+        // underlying address.
+        foreach (idx, ref res; this.resolvers)
+        {
+            assert(peers[idx].host.guessAddressType != TYPE.CNAME,
+                   "Address '" ~ peers[idx].host ~
+                   "' is not an IP address and can't be used for DNS resolution");
+
+            res.address = resolveHost(peers[idx].host);
+            res.address.port = peers[idx].port;
+            res.connection = listenUDP(0, "0.0.0.0",
+                UDPListenOptions.reuseAddress | UDPListenOptions.reusePort);
+            res.connection.connect(res.address);
+        }
+    }
+
+    /***************************************************************************
+
+        Returns the `ResourceRecord` matching `type` associated with `name`
+
+        This low-level function will query the registered resolvers for the
+        records matching `name`. `type` is an optional argument indicating what
+        kind of RR is expected. While it defaults to `ALL`, it is recommended
+        to provide a different value, as many servers might refuse to answer
+        for queries on which they are not authoritative.
+
+        Params:
+          name = The name to resolve
+          type = Type of reecord to query
+
+    ***************************************************************************/
+
+    public ResourceRecord[] query (const(char)[] name, QTYPE type = QTYPE.ALL) @trusted
+    {
+        Message msg;
+        msg.questions ~= Question(Domain.fromString(name), type, QCLASS.IN);
+
+        msg.header.ID = uniform!short;
+        // Don't do the whole recursion dance, assume our resolvers are recursive.
+        msg.header.RD = true;
+        // Default:
+        // QR set to 0 (query), OPCODE set to 0 (Query), AA set to false,
+        // TC set to false, RA set to false, Z set to 0, RCODE set to 0 (NoError)
+        msg.header.QDCOUNT = 1;
+        // Rest is set to 0
+        msg.additionals ~= OPTRR.init.record;
+        msg.header.ARCOUNT = 1;
+
+        ubyte[16384] buffer;
+        foreach (ref peer; this.resolvers)
+        {
+            auto bytes = msg.serializeFull;
+            ubyte[] response;
+            try
+            {
+                peer.connection.send(bytes);
+                response = peer.connection.recv(5.seconds, buffer);
```

Comment:
This will prevent concurrent requests.

Reply:
See https://github.com/bosagora/agora/pull/2790/commits/a7af94f3a38ce0ad3a1c2c30c514b337000fd33b

Reply:
Do we need concurrent requests? I thought this was for backups.

Reply:
Incorporated.